credentials/gcp.go

```
@@ -0,0 +1,171 @@
+	ClientSecret            string   `json:"client_secret"`
+	ProjectID               string   `json:"project_id"`
+	RedirectUris            []string `json:"redirect_uris"`
+	TokenURI                string   `json:"token_uri"`
+}
+
+// DeleteCredential removes a stored credential from the system keystore, and
+// will remove the assigned defualt credential from configuration.
+func (gc GoogleCredential) DeleteCredential(credentialName string) error {
+	return keyring.Delete("scuttle", credentialName+"-gcp")
+}
+
+// SaveCredential marshals the received GoogleToken and stores the resulting json blob in the
+// system keystore for secure storage at rest.
+func (gc GoogleCredential) SaveCredential(credentialName string, credential GoogleToken) error {
+	jsonData, err := json.Marshal(credential)
+	if err != nil {
+		return err
+	}
+	return keyring.Set("scuttle", credentialName+"-gcp", string(jsonData))
+}
+
+// GetCredential returns a decoded GoogleToken from the system keystore. The resulting
+// object is serializeable and should be used in conjunction with google.configFromJSON()
+func (gc GoogleCredential) GetCredential(credentialName string) (GoogleToken, error) {
+	userTokenJSON, err := keyring.Get("scuttle", credentialName+"-gcp")
+	if err != nil {
+		return GoogleToken{}, err
+	}
+	var token GoogleToken
+	err = json.Unmarshal([]byte(userTokenJSON), &token)
+	if err != nil {
+		return GoogleToken{}, err
+	}
+
+	return token, nil
+}
+
+func (gc GoogleCredential) JSON() ([]byte, error) {
+	// This serializer method handles the override from ENV transparently
+	// to the user. We should signal that we're loading from the ENV var
+	// declared credential.
+	external, exists := os.LookupEnv("GOOGLE_APPLICATION_CREDENTIALS")
+	if exists {
+		log.Printf("Using env exported GOOGLE_APPLICATION_CREDENTIALS")
+		f, err := ioutil.ReadFile(external)
+		if err != nil {
+			log.Fatalf("Unable to read GOOGLE_APPLICATION_CREDENTIALS: %v", err)
```

instead of log.Fatal, this should return the error, so the upstream caller can act on the error appropriately (this also makes it easier to test)

you can use https://github.com/pkg/errors to wrap the error too

```go
return []byte{}, errors.Wrap(err, "Unable to read GOOGLE_APPLICATION_CREDENTIALS")
```